File: solutions/python/greedy/monotone-increasing-digits.py
# LeetCode 738 — Monotone Increasing Digits (Medium)
# Category: Greedy · Approach: Greedy from right
# Time: O(d) | Space: O(d)
# Source: https://leetcode.com/problems/monotone-increasing-digits/

def monotone_increasing_digits(n: int) -> int:
    d = list(map(int, str(n)))
    mark = len(d)
    for i in range(len(d) - 1, 0, -1):
        if d[i - 1] > d[i]:
            d[i - 1] -= 1
            mark = i
    for i in range(mark, len(d)):
        d[i] = 9
    return int("".join(map(str, d)))

File: solutions/typescript/greedy/monotone-increasing-digits.ts
// LeetCode 738 — Monotone Increasing Digits (Medium)
// Category: Greedy · Approach: Greedy from right
// Time: O(d) | Space: O(d)
// Source: https://leetcode.com/problems/monotone-increasing-digits/

function monotoneIncreasingDigits(n: number): number {
  const d = String(n).split("").map(Number);
  let mark = d.length;
  for (let i = d.length - 1; i > 0; i--) {
    if (d[i - 1] > d[i]) {
      d[i - 1]--;
      mark = i;
    }
  }
  for (let i = mark; i < d.length; i++) d[i] = 9;
  return Number(d.join(""));
}
